### PR TITLE
Refs #136 - Update dynamic schema on all ZEO clients on change

### LINKED ISSUE #136: SchemaCache is not updated across multiple Plone instances
**How to reproduce:**

1. Start a ZEO cluster with 2 Plone ZEO clients
2. Add Plone Site
3. Go to first Plone instance > Site Setup > Dexterity Content-Types > Page
4. Add new fieldset and new text field and press "Done"
5. Go to the second Plone instance > Site Setup > Dexterity Content-Types > Page
6. Notice the missing custom **fieldset / field**

Plone version: 5.2.2

See also: 
- https://github.com/plone/plone.schemaeditor/issues/81
- https://community.plone.org/t/dynamic-schema-is-not-synced-across-zeo-clients/12923

### PR DESCRIPTION
**Pull request by avoinea**
Fixes #136

### CODE REVIEW

**Comment by avoinea** [comment]
@jenkins-plone-org please run jobs

**Review by jensens — APPROVED** [review]
Simple and elegant to use the `_p_mtime` as part of the generated schema name!

**Comment by jensens** [comment]
@jenkins-plone-org please run jobs

**Comment by jensens** [comment]
@jenkins-plone-org please run jobs

**Comment by avoinea** [comment]
@jensens I'll take a look at the remaining failing tests...

**Comment by avoinea** [comment]
@jensens I don't know what is wrong with this one https://jenkins.plone.org/job/pull-request-5.2/1643/testReport/junit/(root)/(empty)/schemaeditor_txt/ and where it comes from. 

For the rest, I added PRs within [plone.app.dexterity](https://github.com/plone/plone.app.dexterity/pull/314) and [plone.app.contenttypes](https://github.com/plone/plone.app.contenttypes/pull/574)

File: plone/dexterity/schema.py [hunk 1]
```diff
@@ -412,6 +414,7 @@ def __call__(self, name, module):
             if name in self._transient_SCHEMA_CACHE:
                 del self._transient_SCHEMA_CACHE[name]
 
+            log.info("Dynamic schema generated: %s", name)
```
Comment (ale-rt) [inline plone/dexterity/schema.py:417]: I would use `log.debug` or completely remove this line.

**Review by ale-rt — APPROVED** [review]
LGTM, thanks a lot!
Please, consider removing that `log.info` statement added by this PR or to change it with a `log.debug`.

File: plone/dexterity/schema.py [hunk 2]
```diff
@@ -414,7 +414,7 @@ def __call__(self, name, module):
             if name in self._transient_SCHEMA_CACHE:
                 del self._transient_SCHEMA_CACHE[name]
 
-            log.info("Dynamic schema generated: %s", name)
+            log.debug("Dynamic schema generated: %s", name)
```
Comment (ale-rt) [inline plone/dexterity/schema.py:417]: :heart:

**Review by ale-rt — APPROVED** [review]

**Comment by ale-rt** [comment]
@jenkins-plone-org please run jobs

**Comment by ale-rt** [comment]
@jenkins-plone-org please run jobs

**Comment by ale-rt** [comment]
@avoinea The Python 3.7 test failure looks unrelated, but there is a real issue with https://jenkins.plone.org/job/pull-request-5.2/1652/testReport/(root)/(empty)/schemaeditor_txt/

Anyway we are close to have this merged :)

**Comment by avoinea** [comment]
@ale-rt Hmmm:

`plone.app.relationfield-2.0.2-py2.7.egg/plone/app/relationfield/tests/../schemaeditor.txt`

**Comment by avoinea** [comment]
@ale-rt @jensens See https://github.com/plone/plone.app.relationfield/pull/31

**Comment by ale-rt** [comment]
@jenkins-plone-org please run jobs

**Comment by avoinea** [comment]
@ale-rt @jensens How can we have this released also for Plone 4.3.x?

**Comment by ale-rt** [comment]
According to https://github.com/plone/buildout.coredev/blob/a4af25bfde395a999844324b45b9912952a6731e/sources.cfg#L45 4.3 usese the plome.dexterity 2.2.x branch.
Theoretically you can backport the patch there before the 4.3.20 is release, see https://github.com/plone/Products.CMFPlone/issues/3166.
I guess other packages will need fixes like the one involved in this PR.

Sincerely I would save myself from that pain.
Also personally I am not sure I will want to approve it. 
In the past we had even simpler backports to 4.3 that caused major pain.

@mauritsvanrees what do you think about backporting this PR to 4.3?

**Comment by mauritsvanrees** [comment]
Plone 4.3.20 has already been released on [28 August](https://community.plone.org/t/plone-4-3-20-soft-released/12750/11?u=mauritsvanrees). It has been waiting since then for a Universal Installer.

I don't mind doing a release from the `plone.dexterity 2.2.x` branch, but it will never get in an official release.

**Comment by mauritsvanrees** [comment]
BTW, the same is true for Plone 5.1.7: it was released [a week ago](https://community.plone.org/t/plone-5-1-7-soft-released/12875/3?u=mauritsvanrees), but waiting on installer. It is the last release from the Plone 5.1 series, so any updates to the relevant `plone.dexterity` branch, 2.6.x, would not get in an official release.
Again, an individual package release is still possible.

**Comment by tisto** [comment]
@avoinea in plone.restapi we have tests that check for undocumented changes in our auto-generated API docs. This PR causes the auto-generated behavior names to change for every test run:

https://github.com/plone/plone.restapi/pull/1033/checks?check_run_id=1572135872

Therefore it is impossible for us to generate a stable REST API documentation. Is there a way to disable this behavior?

**Comment by avoinea** [comment]
@tisto True. 

We have 2 options I can think of right now:
- Either patch `portalTypeToSchemaName` before the tests,
- or tweak the object `_p_mtime`

Another, elegant solution would be to convert `portalTypeToSchemaName` into an utility to be able to easily override it on need.

**Comment by tisto** [comment]
@avoinea could you look into fixing this in plone.restapi maybe? I don't really have the time for this right now and I ran into this again.

**Pull request merged**